File: deck/base/code/native/kotlin/runtime/eddsa.kt
// Ed25519 signatures over java.security (built into the JDK since 15, no extra dependency). Keys, messages, and
// signatures are raw bytes (ByteArray): a key-pair is the 32-byte private seed and 32-byte public key concatenated
// (64 bytes). To rebuild a key from raw bytes the fixed Ed25519 DER prefixes are prepended (PKCS8 for private, X.509
// for public). Fully qualified (no top-level imports) so the file can be prepended as a runtime prelude. Reached only
// through the public signature API.
object eddsa {
    private const val PKCS8_PREFIX = "302e020100300506032b657004220420"
    private const val SPKI_PREFIX = "302a300506032b6570032100"
    private fun fromHex(hex: String): ByteArray = ByteArray(hex.length / 2) { hex.substring(it * 2, it * 2 + 2).toInt(16).toByte() }
    fun makeKeyPair(): ByteArray {
        val pair = java.security.KeyPairGenerator.getInstance("Ed25519").generateKeyPair()
        val seed = (pair.private as java.security.interfaces.EdECPrivateKey).bytes.get()
        val encoded = pair.public.encoded
        val public = encoded.copyOfRange(encoded.size - 32, encoded.size)
        return seed + public
    }
    fun sign(privateKey: ByteArray, message: ByteArray): ByteArray {
        val key = java.security.KeyFactory.getInstance("Ed25519")
            .generatePrivate(java.security.spec.PKCS8EncodedKeySpec(fromHex(PKCS8_PREFIX) + privateKey))
        val signer = java.security.Signature.getInstance("Ed25519")
        signer.initSign(key)
        signer.update(message)
        return signer.sign()
    }
    fun verify(publicKey: ByteArray, message: ByteArray, signature: ByteArray): Boolean {
        val key = java.security.KeyFactory.getInstance("Ed25519")
            .generatePublic(java.security.spec.X509EncodedKeySpec(fromHex(SPKI_PREFIX) + publicKey))
        val verifier = java.security.Signature.getInstance("Ed25519")
        verifier.initVerify(key)
        verifier.update(message)
        return verifier.verify(signature)
    }
}

File: deck/base/code/native/swift/runtime/eddsa.swift
import Foundation
import CryptoKit

// Ed25519 signatures over CryptoKit. The private key and public key are raw 32-byte representations; a signature is
// 64 bytes. make-key-pair returns the private key and public key concatenated as hex.
enum eddsa {
    private static func fromHex(_ hex: String) -> Data {
        var data = Data(capacity: hex.count / 2)
        var index = hex.startIndex
        while index < hex.endIndex {
            let next = hex.index(index, offsetBy: 2)
            data.append(UInt8(hex[index..<next], radix: 16)!)
            index = next
        }
        return data
    }
    private static func toHex(_ data: Data) -> String { data.map { String(format: "%02x", $0) }.joined() }
    static func makeKeyPair() -> String {
        let signing = Curve25519.Signing.PrivateKey()
        return toHex(signing.rawRepresentation) + toHex(signing.publicKey.rawRepresentation)
    }
    static func sign(_ privateHex: String, _ message: String) -> String {
        let signing = try! Curve25519.Signing.PrivateKey(rawRepresentation: fromHex(privateHex))
        return toHex(try! signing.signature(for: Data(message.utf8)))
    }
    static func verify(_ publicHex: String, _ message: String, _ signatureHex: String) -> Bool {
        guard let key = try? Curve25519.Signing.PublicKey(rawRepresentation: fromHex(publicHex)) else { return false }
        return key.isValidSignature(fromHex(signatureHex), for: Data(message.utf8))
    }
}
